drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1435 @@
+	size_t len;
+
+	if (ep_regs[ep_fifo].ep_status & EP_STATUS_ERROR) {
+		LOG_WRN("endpoint%d error status 0x%02x", ep_idx, ep_regs[ep_fifo].ep_status);
+		return -EINVAL;
+	}
+
+	len = (uint16_t)ff_regs[ep_fifo].ep_rx_fifo_dcnt_lsb +
+	      (((uint16_t)ff_regs[ep_fifo].ep_rx_fifo_dcnt_msb) << 8);
+
+	len = MIN(net_buf_tailroom(buf), len);
+	uint8_t *data_ptr = net_buf_tail(buf);
+
+	for (size_t idx = 0; idx < len; idx++) {
+		data_ptr[idx] = ff_regs[ep_fifo].ep_rx_fifo_data;
+	}
+
+	net_buf_add(buf, len);
+
+	return 0;
+}
+
+static int work_handler_xfer_continue(const struct device *dev, uint8_t ep, struct net_buf *buf)
+{
+	int ret = 0;
+
+	if (USB_EP_DIR_IS_OUT(ep)) {
+		const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+		struct it82xx2_priv_data *priv = udc_get_private(dev);
+		uint8_t ep_fifo = (ep_idx > 0) ? (ep_fifo_res[ep_idx % SHARED_FIFO_NUM]) : 0;
+
+		it82xx2_usb_set_ep_ctrl(dev, ep_idx, EP_READY_ENABLE, true);
+		if (IT8XXX2_IS_EXTEND_ENDPOINT(ep_idx)) {
+			it82xx2_usb_set_ep_ctrl(dev, ep_fifo, EP_READY_ENABLE, true);
+		}
+		if (ep_idx != 0) {
+			priv->fifo_ready[ep_fifo - 1] = true;
+		}
+	} else {
+		ret = it82xx2_xfer_in_data(dev, ep, buf);
+	}
+
+	return ret;
+}
+
+static int work_handler_xfer_next(const struct device *dev, uint8_t ep)
+{
+	struct net_buf *buf;
+
+	buf = udc_buf_peek(dev, ep);
+	if (buf == NULL) {
+		return -ENODATA;
+	}
+
+	return work_handler_xfer_continue(dev, ep, buf);
+}
+
+/*
+ * Allocate buffer and initiate a new control OUT transfer,
+ * use successive buffer descriptor when next is true.
+ */
+static int it82xx2_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+	udc_buf_put(cfg, buf);
+
+	it82xx2_usb_set_ep_ctrl(dev, 0, EP_READY_ENABLE, true);
+
+	return 0;
+}
+
+static bool it82xx2_fake_token(const struct device *dev, uint8_t ep, uint8_t token_type)
+{
+	struct it82xx2_priv_data *priv = udc_get_private(dev);
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	uint8_t ep_fifo;
+	bool is_fake = true;
+
+	if (ep_idx == 0) {
+		switch (token_type) {
+		case DC_IN_TRANS:
+			if (priv->stall_is_sent) {
+				return true;
+			}
+			is_fake = !udc_ctrl_stage_is_data_in(dev) &&
+				  !udc_ctrl_stage_is_status_in(dev) &&
+				  !udc_ctrl_stage_is_no_data(dev);
+			break;
+		case DC_OUTDATA_TRANS:
+			is_fake = !udc_ctrl_stage_is_data_out(dev) &&
+				  !udc_ctrl_stage_is_status_out(dev);
+			break;
+		default:
+			LOG_ERR("Invalid token type");
+			break;
+		}
+	} else {
+		ep_fifo = ep_fifo_res[ep_idx % SHARED_FIFO_NUM];
+
+		if (!priv->fifo_ready[ep_fifo - 1]) {
+			is_fake = true;
+		} else {
+			priv->fifo_ready[ep_fifo - 1] = false;
+			is_fake = false;
+		}
+	}
+
+	return is_fake;
+}
+
+static inline int work_handler_in(const struct device *dev, uint8_t ep);
```

Comment:
Why you don't move the whole function definition here?

Reply:
fixed. thanks.